crates/flowctl/src/pagination.rs

```diff
@@ -0,0 +1,80 @@
+use std::marker::PhantomData;
+
+use page_turner::PageTurner;
+use page_turner::PageTurnerOutput;
+use page_turner::TurnedPage;
+use tonic::async_trait;
+
+use crate::api_exec;
+
+/// A simple wrapper around [`postgrest::Builder`] that lets us keep track of which page
+/// it's currently on. Used in [`page_turner::PageTurner`] to implement pagination.
+pub struct PaginationRequest {
+    builder: postgrest::Builder,
+    page: usize,
+    page_size: usize,
+}
+
+impl PaginationRequest {
+    pub fn new(builder: postgrest::Builder) -> Self {
+        Self {
+            builder,
+            page: 0,
+            page_size: 1000,
+        }
+    }
+
+    fn set_page(mut self, page: usize) -> Self {
+        self.page = page;
+        self.builder = self
+            .builder
+            .range(page * self.page_size, (page + 1) * self.page_size);
+
+        self
+    }
+}
+
+/// A placeholder struct onto which we can implement [`page_turner::PageTurner`].
+/// Normally this would be the API client responsible for actually executing the requests
+/// defined in [`PaginationRequest`], but since a [`postgrest::Builder`] already has
+/// its own client and is responsible for making its own requests, this is empty.
+pub struct PaginationClient<Item>
+where
+    for<'de> Item: serde::Deserialize<'de> + Send + Sync,
+{
+    phantom: PhantomData<fn() -> Item>,
+}
+
+impl<T> PaginationClient<T>
+where
+    for<'de> T: serde::Deserialize<'de> + Send + Sync,
+{
+    pub fn new() -> Self {
+        Self {
+            phantom: PhantomData,
+        }
+    }
+}
+
+#[async_trait]
+impl<Item> PageTurner<PaginationRequest> for PaginationClient<Item>
+where
+    for<'de> Item: serde::Deserialize<'de> + Send + Sync,
+{
+    type PageItem = Item;
+    type PageError = anyhow::Error;
+
+    async fn turn_page(
+        &self,
+        request: PaginationRequest,
+    ) -> PageTurnerOutput<Self, PaginationRequest> {
+        let resp: Vec<Item> = api_exec::<Vec<Item>>(request.builder.clone()).await?;
+
+        if resp.len() == request.page_size {
```

Comment:
Are you familiar with the `Content-Range` header? [PostgREST seems to use it](https://postgrest.org/en/stable/references/api/tables_views.html#limits-and-pagination), although there's some caveats about the total number of items that gets reported, so maybe there's a reason to prefer the heuristic you have here.

Can you verify the behavior of the API when there's a page size of _n_ and exactly _n_ items? We'd send a last request that we'd expect to return a successful response with no items, and I just want to make sure that's what it does.

Reply:
`Content-Range`: Yeeeeah, I saw that, but then also saw that it seems to sometimes return an unknown total count i.e `0-19/*` instead of `0-19/50`, in which case we'd need to fall back to checking the response length anyway, so I figured it wasn't worth complicating things.

Validation:

```
├────────────────────────────────────┼────────────┼─────────────────────────────┼─────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ 0a:1c:78:2a:b7:65:cc:00            │ 28         │ 2023-06-15T18:43:40.514676Z │ 2023-06-15T18:43:48.635952Z │ estuary/marketing/hubspot/source-hubspot                                                           │
├────────────────────────────────────┼────────────┼─────────────────────────────┼─────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────────┤
```

```
page_size: 28
```

Debug output:

```
$ flowctl draft develop
DEBUG flowctl::pagination: Got back a full response, progressing to the next page current_page=0 row_count=28
DEBUG flowctl::pagination: Got back a non-full response or we reached the builder's original limit so we're done current_page=1 row_count=0
Wrote 28 specifications under file:///Users/js/Documents/estuary/flow.yaml.
```

Works as expected!